From ea6e2f7e52a6950bd9b1bcb386336c1d9166d661 Mon Sep 17 00:00:00 2001
From: Andrey Grodzovsky <Andrey.Grodzovsky@amd.com>
Date: Mon, 16 May 2016 13:20:44 -0400
Subject: [PATCH 0487/1722] drm/amd/dal: Fix headless MST crash.

get_connector_for_sink would never retrun NULL in case there
was no matching connector to given sink. Same for get_connector_for_link.

Signed-off-by: Andrey Grodzovsky <Andrey.Grodzovsky@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 .../gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c  | 38 +++++++++++++++-------
 1 file changed, 26 insertions(+), 12 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c
index f85f55f..4173c1f 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c
@@ -142,15 +142,14 @@ static struct amdgpu_connector *get_connector_for_sink(
 	const struct dc_sink *sink)
 {
 	struct drm_connector *connector;
-	struct amdgpu_connector *aconnector = NULL;
 
 	list_for_each_entry(connector, &dev->mode_config.connector_list, head) {
-		aconnector = to_amdgpu_connector(connector);
+                struct amdgpu_connector *aconnector = to_amdgpu_connector(connector);
 		if (aconnector->dc_sink == sink)
-			break;
+                        return aconnector;
 	}
 
-	return aconnector;
+        return NULL;
 }
 #endif
 
@@ -159,15 +158,15 @@ static struct amdgpu_connector *get_connector_for_link(
 	const struct dc_link *link)
 {
 	struct drm_connector *connector;
-	struct amdgpu_connector *aconnector = NULL;
 
 	list_for_each_entry(connector, &dev->mode_config.connector_list, head) {
-		aconnector = to_amdgpu_connector(connector);
+                struct amdgpu_connector *aconnector = to_amdgpu_connector(connector);
 		if (aconnector->dc_link == link)
 			break;
+                        return aconnector;
 	}
 
-	return aconnector;
+        return NULL;
 }
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 3, 0)
@@ -232,7 +231,7 @@ bool dm_helpers_dp_mst_write_payload_allocation_table(
 
 	aconnector = get_connector_for_sink(dev, stream->sink);
 
-	if (!aconnector->mst_port)
+	if (!aconnector || !aconnector->mst_port)
 		return false;
 
 	mst_mgr = &aconnector->mst_port->mst_mgr;
@@ -320,7 +319,7 @@ bool dm_helpers_dp_mst_poll_for_allocation_change_trigger(
 
 	aconnector = get_connector_for_sink(dev, stream->sink);
 
-	if (!aconnector->mst_port)
+	if (!aconnector || !aconnector->mst_port)
 		return false;
 
 	mst_mgr = &aconnector->mst_port->mst_mgr;
@@ -354,11 +353,11 @@ bool dm_helpers_dp_mst_send_payload_allocation(
 
 	aconnector = get_connector_for_sink(dev, stream->sink);
 
-	mst_port = aconnector->port;
-
-	if (!aconnector->mst_port)
+	if (!aconnector || !aconnector->mst_port)
 		return false;
 
+	mst_port = aconnector->port;
+
 	mst_mgr = &aconnector->mst_port->mst_mgr;
 
 	if (!mst_mgr->mst_state)
@@ -434,6 +433,11 @@ bool dm_helpers_dp_mst_start_top_mgr(
 	struct drm_device *dev = adev->ddev;
 	struct amdgpu_connector *aconnector = get_connector_for_link(dev, link);
 
+	if (!aconnector) {
+			DRM_ERROR("Failed to found connector for link!");
+			return false;
+	}
+
 	if (boot) {
 		DRM_INFO("DM_MST: Differing MST start on aconnector: %p [id: %d]\n",
 					aconnector, aconnector->base.base.id);
@@ -458,6 +462,11 @@ void dm_helpers_dp_mst_stop_top_mgr(
 	struct drm_device *dev = adev->ddev;
 	struct amdgpu_connector *aconnector = get_connector_for_link(dev, link);
 
+	if (!aconnector) {
+			DRM_ERROR("Failed to found connector for link!");
+			return;
+	}
+
 	DRM_INFO("DM_MST: stopping TM on aconnector: %p [id: %d]\n",
 			aconnector, aconnector->base.base.id);
 
@@ -561,6 +570,11 @@ void dm_helper_conn_log(struct dc_context *ctx,
 	int size;
 	enum log_minor minor = event;
 
+	if (!aconnector) {
+			DRM_ERROR("Failed to found connector for link!");
+			return;
+	}
+
 	va_start(args, msg);
 
 	sprintf(buffer, "[%s] ", aconnector->base.name);
-- 
2.7.4

